File: web/src/lib/paginacion.ts
// Contrato de paginación de la PWA (spec 2026-09-14): espeja el contrato del
// backend (PeticionPaginada y Pagina<T> de BuildingBlocks.Application) para
// que los listados hablen el mismo idioma que la API.
export interface PeticionPaginada {
  pagina: number;
  tamanoPagina: number;
}

export interface Pagina<T> {
  items: T[];
  total: number;
  numeroPagina: number;
  tamanoPagina: number;
}

export const TAMANO_PAGINA_POR_DEFECTO = 20;

// El backend acota el tamaño a 1–100; la UI solo necesita saber cuántas
// páginas hay para el total y el tamaño vigentes.
export function contarPaginas(total: number, tamanoPagina: number): number {
  return Math.max(1, Math.ceil(total / Math.max(tamanoPagina, 1)));
}

// Query string común de los listados filtrables: página, tamaño y solo los
// filtros con valor (vacío significa «sin filtrar»).
export function consultaPaginada(
  peticion: PeticionPaginada,
  filtros: Record<string, string | undefined> = {},
): string {
  const parametros = new URLSearchParams({
    pagina: String(peticion.pagina),
    tamanoPagina: String(peticion.tamanoPagina),
  });
  for (const [clave, valor] of Object.entries(filtros)) {
    if (valor) parametros.set(clave, valor);
  }
  return parametros.toString();
}

File: Icarus/src/BuildingBlocks/Icarus.BuildingBlocks.Application/Paginacion.cs
namespace Icarus.BuildingBlocks.Application;

// Contrato único de paginación (spec 2026-09-14). Los límites son los que ya
// usaba ListarPedidosCaisyValidator: no se inventaron acá, se unificaron.
// La normalización vive en el record y no en cada handler para que una página
// inválida no llegue nunca a SQL Server como un OFFSET negativo.
public sealed record PeticionPaginada(int Pagina = 1, int TamanoPagina = PeticionPaginada.TamanoPorDefecto)
{
    public const int TamanoPorDefecto = 20;
    public const int TamanoMaximo = 100;

    public int PaginaNormalizada => Math.Max(Pagina, 1);

    public int TamanoNormalizado => TamanoPagina switch
    {
        < 1 => TamanoPorDefecto,
        > TamanoMaximo => TamanoMaximo,
        _ => TamanoPagina,
    };

    public int Salto => (PaginaNormalizada - 1) * TamanoNormalizado;
}

// Total es el conteo SIN paginar: la UI lo necesita para saber cuántas
// páginas hay, y es lo único que obliga a una segunda consulta.
public sealed record Pagina<T>(
    IReadOnlyList<T> Items, int Total, int NumeroPagina, int TamanoPagina);
